src/lib/storage/dictionary_column.hpp

```diff
@@ -25,48 +31,107 @@ class DictionaryColumn : public BaseColumn {
   /**
    * Creates a Dictionary column from a given value column.
    */
-  explicit DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column);
+  explicit DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column) { _compress_values(base_column); }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionaryColumn in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  const AllTypeVariant operator[](const size_t i) const override;
+  const AllTypeVariant operator[](const size_t i) const override { return AllTypeVariant{get(i)}; }
 
   // return the value at a certain position.
-  const T get(const size_t i) const;
+  const T get(const size_t i) const { return _dictionary->at(_attribute_vector->get(i)); }
 
   // dictionary columns are immutable
-  void append(const AllTypeVariant&) override;
+  void append(const AllTypeVariant&) override { throw std::runtime_error("DictionaryColumns are immutable"); };
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const {
+    return std::dynamic_pointer_cast<const BaseAttributeVector>(_attribute_vector);
+  }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    auto iterator = std::lower_bound(_dictionary->cbegin(), _dictionary->cend(), value);
+    if (iterator == _dictionary->cend()) {
+      return INVALID_VALUE_ID;
+    }
+
+    return ValueID{static_cast<ValueID::base_type>(iterator - _dictionary->cbegin())};
```

Comment:
Unfortunately it is necessary (at least with GCC 7.2), as the expression returns an `iterator::difference_type` (which is `long int` on my platform) and `ValueID` is a `uint32_t`, so this would be a narrowing conversion.